cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java

```diff
@@ -126,45 +155,121 @@
     this.streamCoordinatorClient = streamCoordinatorClient;
     this.privilegesManager = privilegesManager;
     this.authorizationEnforcer = authorizationEnforcer;
+    // TODO make maximum size and expire after write configurable?
+    this.appInjectors = CacheBuilder.newBuilder()
+      .maximumSize(10)
+      .expireAfterWrite(15, TimeUnit.MINUTES)
+      .removalListener(new RemovalListener<ApplicationId, Injector>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<ApplicationId, Injector> notification) {
+          Injector injector = notification.getValue();
+          if (injector != null) {
+            PreviewRuntimeService service = injector.getInstance(PreviewRuntimeService.class);
+            service.stopAndWait();
+          }
+          ApplicationId application = notification.getKey();
+          if (application == null) {
```

Comment:
This will not be null in our case, it will only be null if we use `weakKeys()`. However `notification.getKey()` is nullable so checking `null` for not having warning.